### PR TITLE
OpenSSL 3.0 support

### PR DESCRIPTION
**Pull request by jschmidb**
This pull request provides support for OpenSSL 3.0 in libica, but also some bug fixes.

### CODE REVIEW

File: src/s390_ecc.c [hunk 1]
```diff
@@ -817,48 +968,48 @@ unsigned int ecdh_hw(ica_adapter_handle_t adapter_handle,
...
 
 #ifdef ICA_FIPS
 	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
-		return EACCES;
+	return EACCES;
 #endif /* ICA_FIPS */
 
-	if (!is_supported_openssl_curve(privkey_A->nid))
-		return EINVAL;
+	if (!is_supported_openssl_curve(privkey_A->nid)) {
```
Comment (ifranzki) [inline src/s390_ecc.c:984]: no { ... } needed for a one liner body (same applies to other places)

File: src/openssl3-fips.cnf.in [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+[openssl_init]
+providers = provider_sect
+
+[provider_sect]
+fips = fips_sect
+base = base_sect
+
+[base_sect]
+activate=1
```
Comment (ifranzki) [inline src/openssl3-fips.cnf.in:13]: I would add a new line after the last line.

File: test/rsa_keygen_test.c [hunk 3]
```diff
@@ -250,9 +250,18 @@ int main(int argc, char **argv)
 	} else {
 		printf("RSA keygen (%u) tests failed: %u errors.",
 		    key_bit_length, rc_test);
-		if (FIPS_mode())
+#ifdef ICA_FIPS
+		if (ica_fips_status()) {
 			printf(" (Parameters might be non FIPS conformant.)");
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
```
Comment (ifranzki) [inline test/rsa_keygen_test.c:256]: are you using OPENSSL_VERSION_NUMBER because OPENSSL_VERSION_PREREQ is not available here?

File: src/fips.c [hunk 4]
```diff
@@ -51,6 +68,8 @@ static const char hmackey[] =
 
 int fips;
 
+#define LIBICA_FIPS_CONFIG		"/etc/libica/openssl3-fips.cnf"
```
Comment (juergenchrist) [inline src/fips.c:71]: I would not hard-code `/etc` here.  That is usually configurable as `sysconfdir` and might default to `/usr/local/etc`.

File: src/include/s390_aes.h [hunk 5]
```diff
@@ -308,12 +312,15 @@ static inline int s390_aes_ecb_sw(unsigned int function_code,
 	unsigned int direction;
 	unsigned int key_size = (function_code & 0x0f) *
 				sizeof(ica_aes_key_single_t);
+	int rc = 0;
 
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/include/s390_aes.h:318]: For OpenSSL 3.0 this simplifies to `(fips & ICA_FIPS_MODE) && (!fips)` which is always false.

File: src/include/s390_aes.h [hunk 6]
```diff
@@ -374,12 +382,15 @@ static inline int s390_aes_cbc_sw(unsigned int function_code,
...
```
Comment (juergenchrist) [inline src/include/s390_aes.h:388]: Always false.  See above.

File: src/include/s390_des.h [hunk 7]
```diff
@@ -52,11 +56,15 @@ static inline int s390_des_ecb_sw(unsigned int function_code, unsigned long inpu
 		    const unsigned char *input_data, const unsigned char *keys,
 		    unsigned char *output_data)
 {
+	int rc = 0;
+
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/include/s390_des.h:62]: Always false.  See above.

File: src/include/s390_des.h [hunk 8]
```diff
@@ -154,14 +163,19 @@ static inline int s390_des_cbc_sw(unsigned int function_code,
 			   const unsigned char *input_data, unsigned char *iv,
 			   const unsigned char *keys, unsigned char *output_data)
 {
+	int rc = 0;
+
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/include/s390_des.h:169]: Always false.  See above.

File: src/s390_rsa.c [hunk 9]
```diff
@@ -54,11 +62,12 @@ static unsigned int mod_expo_sw(int arg_length, char *arg, int exp_length,
 				int *res_length, char *res, BN_CTX *ctx);
 #endif /* NO_SW_FALLBACKS */
 
+#if !OPENSSL_VERSION_PREREQ(3, 0)
 RSA* rsa_key_generate(unsigned int modulus_bit_length,
 		      unsigned long *public_exponent)
 {
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/s390_rsa.c:70]: Always false.  See above.

File: src/s390_rsa.c [hunk 10]
```diff
@@ -128,26 +183,50 @@ unsigned int rsa_key_generate_mod_expo(ica_adapter_handle_t deviceHandle,
 				       ica_rsa_key_mod_expo_t *private_key)
 {
 	(void)deviceHandle;	/* suppress unused param warning */
+	int rc = 0;
 
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/s390_rsa.c:189]: Always false.  See above.

File: src/s390_rsa.c [hunk 11]
```diff
@@ -200,32 +286,56 @@ unsigned int rsa_key_generate_crt(ica_adapter_handle_t deviceHandle,
 				  ica_rsa_key_crt_t *private_key)
 {
 	(void)deviceHandle;	/* suppress unused param warning */
+	int rc = 0;
 
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/s390_rsa.c:292]: Always false.  See above.

File: src/s390_rsa.c [hunk 12]
```diff
@@ -311,7 +427,7 @@ unsigned int rsa_mod_mult_sw(ica_rsa_modmult_t *pMul)
 	BN_CTX *ctx = NULL;
 
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/s390_rsa.c:430]: Always false.  See above.

File: src/s390_rsa.c [hunk 13]
```diff
@@ -347,7 +463,7 @@ static unsigned int mod_mul_sw(int fc_1_length, char *fc1, int fc_2_length,
 	BIGNUM *b_res = NULL;
 
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/s390_rsa.c:466]: Always false.  See above.

File: src/s390_rsa.c [hunk 14]
```diff
@@ -420,7 +536,7 @@ unsigned int rsa_mod_expo_sw(ica_rsa_modexpo_t *pMex)
 	BN_CTX *ctx = NULL;
 
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/s390_rsa.c:539]: Always false.  See above.

File: src/s390_rsa.c [hunk 15]
```diff
@@ -486,7 +602,7 @@ static unsigned int mod_expo_sw(int arg_length, char *arg, int exp_length,
 	int mod_expo_rc = 1;
 
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/s390_rsa.c:605]: Always false.  See above.

File: src/s390_rsa.c [hunk 16]
```diff
@@ -573,7 +689,7 @@ unsigned int rsa_crt_sw(ica_rsa_modexpo_crt_t * pCrt)
 	BN_CTX *ctx = NULL;
 
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/s390_rsa.c:692]: Always false.  See above.

File: src/s390_rsa.c [hunk 17]
```diff
@@ -720,7 +836,7 @@ static unsigned int mod_sw(int arg_length, char *arg, int mod_length,
 	BIGNUM *b_res = NULL;
 
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/s390_rsa.c:839]: Always false.  See above.

File: src/s390_rsa.c [hunk 18]
```diff
@@ -801,7 +917,7 @@ static unsigned int mod_sub_sw(int min_length, char *minu, int sub_length,
 	int min_size, sub_size, dif_size;
 
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/s390_rsa.c:920]: Always false.  See above.

File: src/s390_rsa.c [hunk 19]
```diff
@@ -907,7 +1023,7 @@ static unsigned int add_sw(int aug_length, char *aug, int add_length,
 	BN_CTX_start(ctx);
 
 #ifdef ICA_FIPS
-	if ((fips & ICA_FIPS_MODE) && (!FIPS_mode()))
+	if ((fips & ICA_FIPS_MODE) && (!openssl_in_fips_mode()))
```
Comment (juergenchrist) [inline src/s390_rsa.c:1026]: Always false.  See above.

File: src/s390_rsa.c [hunk 20]
```diff
@@ -976,7 +1092,7 @@ static unsigned int mul_sw(int fc_1_length, char *fc1, int fc_2_length,
...
```
Comment (juergenchrist) [inline src/s390_rsa.c:1095]: Always false.  See above.

**Review by juergenchrist — CHANGES_REQUESTED** [review]

**Pull request merged**